### PR TITLE
introduce simplify_exprt::resultt

### PR DESCRIPTION
**Pull request by kroening**
This is follow-up on #2939 and #2946, and the biggest change to the simplifier since its existence.

It attempts to make the implementation of individual simplification steps more type safe, as opposed to using `exprt`.

The alternative considered was to return `optionalt<exprt>`, which has the disadvantage that it doesn't "chain", i.e., given two transformations `f` and `g` one cannot write `f(g(...))`.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
Could the commit message please include a body - which might actually just what's being said in the PR description?

**Review by tautschnig — COMMENTED** [review]
The commit in itself looks promising, but also it's not easy to judge without seeing it in action.

**Comment by kroening** [comment]
@tautschnig Now with non-trivial example.

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -2405,139 +2405,139 @@ bool simplify_exprt::simplify_node_preorder(exprt &expr)
 bool simplify_exprt::simplify_node(exprt &expr)
 {
   if(!expr.has_operands())
-    return true;
+    return true; // no change
 
-  // #define DEBUGX
+    // #define DEBUGX
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp:2371]: Except for this seemingly wrong/unnecessary change: could we please just have this commit merged as soon as possible? It seems like a big improvement towards readability without any immediate dependence on the other changes.

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -272,91 +272,79 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+  if(expr.op().id() == ID_infinity)
   {
     typet new_type=expr.type();
-    exprt tmp;
-    tmp.swap(expr.op0());
+    exprt tmp = expr.op();
     tmp.type()=new_type;
-    expr.swap(tmp);
-    return false;
+    return tmp;
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Do we need `std::move(tmp)` here?

Comment (tautschnig) [reply src/util/simplify_expr.cpp]: Actually, don't we need `return changed(std::move(tmp));` here?

File: src/util/simplify_expr.cpp [hunk 3]
```diff
@@ -272,91 +272,79 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
     binary_relation_exprt inequality(
-      expr.op0(),
+      expr.op(),
       op_type.id() == ID_floatbv ? ID_ieee_float_notequal : ID_notequal,
       from_integer(0, op_type));
     inequality.add_source_location()=expr.source_location();
     simplify_node(inequality);
-    expr.swap(inequality);
-    return false;
+    return inequality;
```
Comment (tautschnig) [inline src/util/simplify_expr.cpp]: Unless I'm mistaken this needs a lot more use of `std::move` at about any return statement?

**Comment by kroening** [comment]
The std::move are added!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 23bfef15f033d2dea9009451e9ead3e2dc923a72).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114390981

**Review by tautschnig — APPROVED** [review]
